File: src/com/chunkmapper/writer/LevelDat.java
package com.chunkmapper.writer;

import java.io.*;

import com.chunkmapper.Point;
import com.chunkmapper.admin.MyLogger;
import com.chunkmapper.gui.dialog.NewMapDialog.GameMode;
import com.chunkmapper.gui.dialog.NewMapDialog.NewGameInfo;
import com.chunkmapper.nbt.ByteTag;
import com.chunkmapper.nbt.CompoundTag;
import com.chunkmapper.nbt.DoubleTag;
import com.chunkmapper.nbt.FloatTag;
import com.chunkmapper.nbt.ListTag;
import com.chunkmapper.nbt.NbtIo;

public class LevelDat {
	private final CompoundTag data;
	public final File store;

	public LevelDat(File loadedLevelDatFile, NewGameInfo newGameInfo) throws IOException {
		this.store = loadedLevelDatFile;
		if (store.exists()) {
			InputStream in = new BufferedInputStream(new FileInputStream(store));
			data = NbtIo.readCompressed(in);
			in.close();
		} else {
			if (newGameInfo == null) {
				File parentFile = loadedLevelDatFile.getParentFile();
				newGameInfo = new NewGameInfo(parentFile.getName());
			}
			
			this.data = getRootTag(newGameInfo);
		}
	}
	public void save() {
		File parent = store.getParentFile();
		parent.mkdirs();
		try {
			OutputStream out = new BufferedOutputStream(new FileOutputStream(store));
			NbtIo.writeCompressed(data, out);
			out.close();
		} catch (IOException e) {
			MyLogger.LOGGER.severe(MyLogger.printException(e));
		}
	}
	public void setPlayerPosition(double x, double y, double z) {
		MyLogger.LOGGER.info("setting player position to " + x + ", " + y + ", " + z);
		CompoundTag Data = data.getCompound("Data");
		Data.putInt("SpawnX", (int) x);
		Data.putInt("SpawnY", (int) y);
		Data.putInt("SpawnZ", (int) z);
		CompoundTag Player = Data.getCompound("Player");
		Player.putInt("SpawnX", (int) x);
		Player.putInt("SpawnY", (int) y);
		Player.putInt("SpawnZ", (int) z);

		ListTag<DoubleTag> Pos = new ListTag<DoubleTag>("Pos");
		Pos.add(new DoubleTag("", x));
		Pos.add(new DoubleTag("", y));
		Pos.add(new DoubleTag("", z));
		Player.put("Pos", Pos);
	}
	public Point getPlayerPosition() {
		CompoundTag Data = data.getCompound("Data");
		CompoundTag Player = Data.getCompound("Player");
		ListTag<DoubleTag> Pos = (ListTag<DoubleTag>) Player.getList("Pos");
		MyLogger.LOGGER.info(store.toString());
		MyLogger.LOGGER.info("***Pos***");
		for (int i = 0; i < 3; i++) {
			MyLogger.LOGGER.info(Pos.get(i).data + "");
		}
		MyLogger.LOGGER.info("***");
		double x = Pos.get(0).data;
		double z = Pos.get(2).data;
		return new Point((int) x, (int) z);
	}
	public void setName(String name) {
		MyLogger.LOGGER.info("setting game name to " + name);
		CompoundTag Data = data.getCompound("Data");
		Data.putString("LevelName", name);
	}
	public String getGameName() {
		CompoundTag Data = data.getCompound("Data");
		return Data.getString("LevelName");
	}
	private static CompoundTag getRootTag(NewGameInfo newGameInfo) {
		CompoundTag root1 = new CompoundTag();

		CompoundTag Data1 = new CompoundTag();
		Data1.putByte("thundering", (byte) 0);
		Data1.putLong("DayTime", 1456L);
		Data1.putLong("LastPlayed", System.currentTimeMillis());

		CompoundTag Player1 = new CompoundTag();
		Player1.putInt("SelectedItemSlot", 0);

		ListTag<FloatTag> DropChances1 = new ListTag<FloatTag>();
		DropChances1.add(new FloatTag("", (float) 0.085));
		DropChances1.add(new FloatTag("", (float) 0.085));
		DropChances1.add(new FloatTag("", (float) 0.085));
		DropChances1.add(new FloatTag("", (float) 0.085));
		DropChances1.add(new FloatTag("", (float) 0.085));
		Player1.put("DropChances", DropChances1);

		Player1.putLong("UUIDLeast", -4690821792555694401L);

		ListTag<DoubleTag> Motion1 = new ListTag<DoubleTag>();
		Motion1.add(new DoubleTag("", 0.0));
		Motion1.add(new DoubleTag("", 0.0));
		Motion1.add(new DoubleTag("", 0.0));
		Player1.put("Motion", Motion1);

		Player1.putString("CustomName", "");
		Player1.putFloat("foodExhaustionLevel", (float) 0.0);
		Player1.putInt("foodTickTimer", 0);
		Player1.putInt("XpLevel", 0);
		Player1.putShort("Health", (short) 20);

		ListTag<ByteTag> Inventory1 = new ListTag<ByteTag>();
		Player1.put("Inventory", Inventory1);

		Player1.putByte("CustomNameVisible", (byte) 0);
		Player1.putShort("AttackTime", (short) 0);
		Player1.putByte("Sleeping", (byte) 0);
		Player1.putShort("Fire", (short) -20);
		//Player1.putInt("playerGameType", 1);
		//replaced, see below
		Player1.putInt("foodLevel", 20);
		Player1.putInt("Score", 0);
		Player1.putByte("Invulnerable", (byte) 0);
		Player1.putShort("DeathTime", (short) 0);

		ListTag<ByteTag> EnderItems1 = new ListTag<ByteTag>();
		Player1.put("EnderItems", EnderItems1);

		Player1.putFloat("XpP", (float) 0.0);
		Player1.putShort("SleepTimer", (short) 0);

		ListTag<CompoundTag> Equipment1 = new ListTag<CompoundTag>();

		CompoundTag anon21 = new CompoundTag();
		Equipment1.add(anon21);


		CompoundTag anon24 = new CompoundTag();
		Equipment1.add(anon24);


		CompoundTag anon27 = new CompoundTag();
		Equipment1.add(anon27);


		CompoundTag anon30 = new CompoundTag();
		Equipment1.add(anon30);


		CompoundTag anon33 = new CompoundTag();
		Equipment1.add(anon33);

		Player1.put("Equipment", Equipment1);

		Player1.putByte("OnGround", (byte) 0);
		Player1.putShort("HurtTime", (short) 0);
		Player1.putLong("UUIDMost", 8633419034662685699L);
		Player1.putInt("Dimension", 0);
		Player1.putShort("Air", (short) 300);
		Player1.putByte("CanPickUpLoot", (byte) 0);

		ListTag<DoubleTag> Pos1 = new ListTag<DoubleTag>();
		Pos1.add(new DoubleTag("", 0));
		Pos1.add(new DoubleTag("", 255));
		Pos1.add(new DoubleTag("", 0));
		Player1.put("Pos", Pos1);

		Player1.putInt("PortalCooldown", 0);
		Player1.putFloat("foodSaturationLevel", (float) 5.0);

//	     walkSpeed: The walking speed, always 0.1.
//	     flySpeed: The flying speed, always 0.05.
//	     mayfly: 1 or 0 (true/false) - true if the player can fly.
//	     flying: 1 or 0 (true/false) - true if the player is currently flying.
//	     invulnerable: 1 or 0 (true/false) - true if the player is immune to all damage and harmful effects except for void damage. (damage caused by the /kill command is void damage)
//	     mayBuild: 1 or 0 (true/false) - true if the player can place and destroy blocks.
//	     instabuild: 1 or 0 (true/false) - true if the player can instantly destroy blocks.
		
		byte playerByte = (byte) (newGameInfo.gameMode == GameMode.Creative_Mode ? 1 : 0);
		
		CompoundTag abilities1 = new CompoundTag();
		abilities1.putByte("flying", (byte) playerByte);
		abilities1.putByte("mayfly", (byte) playerByte);
		abilities1.putByte("instabuild", (byte) playerByte);
		abilities1.putByte("invulnerable", (byte) playerByte);
		abilities1.putByte("mayBuild", (byte) 1);
		abilities1.putFloat("flySpeed", (float) 0.05);
		abilities1.putFloat("walkSpeed", (float) 0.1);
		Player1.put("abilities", abilities1);

		Player1.putByte("PersistenceRequired", (byte) 0);
		Player1.putFloat("FallDistance", (float) 0.0);
		Player1.putInt("XpTotal", 0);

		ListTag<FloatTag> Rotation1 = new ListTag<FloatTag>();
		Rotation1.add(new FloatTag("", (float) 277.34998));
		Rotation1.add(new FloatTag("", (float) 13.950021));
		Player1.put("Rotation", Rotation1);

		Data1.put("Player", Player1);

		Data1.putByte("initialized", (byte) 1);
		Data1.putLong("RandomSeed", 6329512828970730748L);
		
		//0 is Survival Mode, 1 is Creative Mode, 2 is Adventure Mode
		int gameType;
		switch(newGameInfo.gameMode) {
		case Survival_Mode:
			gameType = 0; break;
		case Creative_Mode:
			gameType = 1; break;
		default:
			gameType = 2; break;
		}
		
		byte difficulty;
		switch(newGameInfo.difficulty) {
		case Peaceful:
			difficulty = 0; break;
		case Easy:
			difficulty = 1; break;
		case Normal:
			difficulty = 2; break;
		default:
			difficulty = 3; break;
		}
		byte difficultyLocked = (byte) (newGameInfo.gameMode == GameMode.Hardcore_Mode ? 1 : 0);
		
		Data1.putInt("GameType", gameType);
		//and also
		Player1.putInt("playerGameType", gameType);
		
		Data1.putByte("Difficulty", difficulty);
		Data1.putByte("DifficultyLocked", difficultyLocked);
		Data1.putByte("MapFeatures", (byte) 0);
		Data1.putInt("version", 19133);
		byte allowCommandsByte = (byte) (newGameInfo.hasCheats ? 1 : 0);
		Data1.putByte("allowCommands", allowCommandsByte);
		Data1.putLong("Time", 1456L);
		Data1.putByte("raining", (byte) 0);
		Data1.putInt("SpawnX", -1354);
		Data1.putInt("thunderTime", 48369);
		Data1.putInt("SpawnY", 4);
		byte hardcoreByte = (byte) (newGameInfo.gameMode == GameMode.Hardcore_Mode ? 1 : 0);
		Data1.putByte("hardcore", hardcoreByte);
		Data1.putInt("SpawnZ", 307);
		Data1.putString("LevelName", "base");
		Data1.putString("generatorOptions", "2;7,12,8,8;1;village");
		Data1.putLong("SizeOnDisk", 0L);
		Data1.putString("generatorName", "flat");

		CompoundTag GameRules1 = new CompoundTag();
		GameRules1.putString("doFireTick", "true");
		GameRules1.putString("doMobLoot", "true");
		GameRules1.putString("mobGriefing", "true");
		GameRules1.putString("doMobSpawning", "false");
		GameRules1.putString("doTileDrops", "true");
		GameRules1.putString("keepInventory", "false");
		GameRules1.putString("commandBlockOutput", "true");
		Data1.put("GameRules", GameRules1);

		Data1.putInt("generatorVersion", 0);
		Data1.putInt("rainTime", 135506);
		root1.put("Data", Data1);
		return root1;

	}
	public static LevelDat getFromGameFolder(File gameFolder) throws IOException {
		return new LevelDat(new File(gameFolder, "level.dat"), null);
	}
	public void setPlayerPosition(double lat, double lon, Point rootPoint) {
		double x = lon * 3600 - rootPoint.x * 512;
		double y = 250;
		double z = lat * 3600 - rootPoint.z * 512;
		this.setPlayerPosition(x, y, z);
	}

}

File: chunkmapperResources/chunkmapper/level_dat.clj
(ns chunkmapper.level-dat
  (:require
    [chunkmapper.nbt :as nbt]))

(defn level-dat []
  (nbt/->nbt
    `{:Data
      {:BorderWarningBlocks [:double 5.0],
       :BorderCenterZ [:double 0.0],
       :BorderDamagePerBlock [:double 0.2],
       :SpawnX [:int 64],
       :WanderingTraderSpawnDelay [:int 24000],
       :initialized [:byte 1],
       :BorderSizeLerpTarget [:double 6.0E7],
       :BorderWarningTime [:double 15.0],
       :WorldGenSettings
       {:bonus_chest [:byte 0],
        :seed [:long 699818219905139133],
        :generate_features [:byte 1],
        :dimensions
        {:minecraft:overworld
         {:generator
          {:settings "minecraft:overworld",
           :seed [:long 699818219905139133],
           :biome_source
           {:seed [:long 699818219905139133],
            :large_biomes [:byte 0],
            :type "minecraft:vanilla_layered"},
           :type "minecraft:noise"},
          :type "minecraft:overworld"},
         :minecraft:the_nether
         {:generator
          {:settings "minecraft:nether",
           :seed [:long 699818219905139133],
           :biome_source
           {:seed [:long 699818219905139133],
            :preset "minecraft:nether",
            :type "minecraft:multi_noise"},
           :type "minecraft:noise"},
          :type "minecraft:the_nether"},
         :minecraft:the_end
         {:generator
          {:settings "minecraft:end",
           :seed [:long 699818219905139133],
           :biome_source
           {:seed [:long 699818219905139133], :type "minecraft:the_end"},
           :type "minecraft:noise"},
          :type "minecraft:the_end"}}},
       :clearWeatherTime [:int 0],
       :WanderingTraderSpawnChance [:int 25],
       :LastPlayed [:long 1612553808898],
       :SpawnY [:int 70],
       :DataVersion [:int 2586],
       :SpawnZ [:int 80],
       :BorderSafeZone [:double 5.0],
       :GameRules
       {:commandBlockOutput "true",
        :doMobLoot "true",
        :sendCommandFeedback "true",
        :spectatorsGenerateChunks "true",
        :fallDamage "true",
        :maxCommandChainLength "65536",
        :fireDamage "true",
        :drowningDamage "true",
        :doImmediateRespawn "false",
        :reducedDebugInfo "false",
        :maxEntityCramming "24",
        :doFireTick "true",
        :doEntityDrops "true",
        :spawnRadius "10",
        :keepInventory "false",
        :universalAnger "false",
        :forgiveDeadPlayers "true",
        :doDaylightCycle "true",
        :mobGriefing "true",
        :randomTickSpeed "3",
        :doMobSpawning "true",
        :doTraderSpawning "true",
        :disableElytraMovementCheck "false",
        :showDeathMessages "true",
        :doInsomnia "true",
        :doTileDrops "true",
        :naturalRegeneration "true",
        :logAdminCommands "true",
        :doPatrolSpawning "true",
        :doWeatherCycle "true",
        :disableRaids "false",
        :doLimitedCrafting "false",
        :announceAdvancements "true"},
       :BorderSizeLerpTime [:long 0],
       :LevelName "New World",
       :thundering [:byte 0],
       :allowCommands [:byte 0],
       :ServerBrands ("vanilla"),
       :BorderSize [:double 6.0E7],
       :GameType [:int 0],
       :CustomBossEvents {},
       :rainTime [:int 64286],
       :DifficultyLocked [:byte 0],
       :thunderTime [:int 92335],
       :SpawnAngle [:float 0.0],
       :DataPacks {:Enabled ("vanilla"), :Disabled ()},
       :Time [:long 157],
       :BorderCenterX [:double 0.0],
       :Player
       {:foodSaturationLevel [:float 5.0],
        :SelectedItemSlot [:int 0],
        :foodLevel [:int 20],
        :Dimension "minecraft:overworld",
        :HurtTime [:short 0],
        :PortalCooldown [:int 0],
        :Score [:int 0],
        :Rotation ([:float 0.0] [:float 0.0]),
        :foodTickTimer [:int 0],
        :XpP [:float 0.0],
        :Motion ([:double 0.0] [:double -0.0784000015258789] [:double 0.0]),
        :DataVersion [:int 2586],
        :Pos ([:double 74.5] [:double 70.0] [:double 76.5]),
        :abilities
        {:invulnerable [:byte 0],
         :mayfly [:byte 0],
         :instabuild [:byte 0],
         :walkSpeed [:float 0.1],
         :mayBuild [:byte 1],
         :flying [:byte 0],
         :flySpeed [:float 0.05]},
        :foodExhaustionLevel [:float 0.0],
        :Fire [:short -20],
        :FallFlying [:byte 0],
        :seenCredits [:byte 0],
        :playerGameType [:int 0],
        :Invulnerable [:byte 0],
        :EnderItems (),
        :FallDistance [:float 0.0],
        :DeathTime [:short 0],
        :XpTotal [:int 0],
        :XpSeed [:int 0],
        :HurtByTimestamp [:int 0],
        :AbsorptionAmount [:float 0.0],
        :OnGround [:byte 1],
        :Air [:short 300],
        :Inventory (),
        :recipeBook
        {:isBlastingFurnaceGuiOpen [:byte 0],
         :isSmokerGuiOpen [:byte 0],
         :toBeDisplayed (),
         :isBlastingFurnaceFilteringCraftable [:byte 0],
         :recipes (),
         :isFurnaceGuiOpen [:byte 0],
         :isFilteringCraftable [:byte 0],
         :isSmokerFilteringCraftable [:byte 0],
         :isFurnaceFilteringCraftable [:byte 0],
         :isGuiOpen [:byte 0]},
        :UUID [:int-array (-1177994134 -965261488 -2069588330 -1020369339)],
        :SleepTimer [:short 0],
        :Attributes
        ({:Base [:double 0.10000000149011612],
          :Name "minecraft:generic.movement_speed"}),
        :Brain {:memories {}},
        :previousPlayerGameType [:int -1],
        :Health [:float 20.0],
        :XpLevel [:int 0]},
       :version [:int 19133],
       :DayTime [:long 157],
       :Version {:Snapshot [:byte 0], :Id [:int 2586], :Name "1.16.5"},
       :WasModded [:byte 0],
       :Difficulty [:byte 2],
       :hardcore [:byte 0],
       :raining [:byte 0],
       :DragonFight
       {:Gateways
        ([:int 9]
         [:int 13]
         [:int 3]
         [:int 11]
         [:int 16]
         [:int 7]
         [:int 6]
         [:int 12]
         [:int 14]
         [:int 18]
         [:int 0]
         [:int 10]
         [:int 8]
         [:int 15]
         [:int 4]
         [:int 5]
         [:int 1]
         [:int 19]
         [:int 2]
         [:int 17]),
        :DragonKilled [:byte 1],
        :PreviouslyKilled [:byte 1]},
       :ScheduledEvents ()}}))
